Chain configuration of ldProgram

erikd opened the pull request:
Cabal currently adds the `-r` (relocatable) flag to some invocations of `ldProgram`, but `lld` (commonly used on WIndows) does not understand the `-r` flag. This PR just detects if `ldProgram` accepts `-r` and conditionally adds it, only if it is supported.

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.

Bonus points for added automated tests!

andreabedini committed:
Partially revert #8071

erikd committed:
Fix configuation of ldProgram

Standard GNU `ld` ues `--relocatable` while `ld.gold` uses a `-relocatable`
flag (with a single `-`). Code will now detect both versions.

erikd committed:
Chain configuration of ldProgram

`ldProgram` gets configured in two places, a seemingly default and
a GHC specific version. The later needs to be updated so that it
first calls the default configuration and then the new GHC version.

andreabedini reviewed (approved):
Much gried was caused by `Distribution.Simple.GHC.Internal` overriding `programPostConf` set in `Distribution.Simple.Program.Builtin`.
I cannot shake the feeing this is an area that could be improved but this is ok for now IMHO.

Mikolaj commented:
Is there an issue ticket connected to this PR? If not, could you add some motivation and explanation to the PR description? Thank you.

andreabedini commented:
@erikd I understand this is a part of a series of patches towards a goal but @Mikolaj is right in pointing out it's not clear where this goal is explicitated :)

I found these links
https://github.com/haskell/cabal/pull/9236
https://github.com/haskell/cabal/pull/9229
I think something was mentioned in https://github.com/haskell/cabal/pull/9226
Then there's https://github.com/input-output-hk/devx-aux/issues/143

Could you write a summary of what you are aiming for?

erikd commented:
@Mikolaj @andreabedini The main issue is that the `lld` linker (which I think is most commonly used on Windows) does not understand the `-r` flag. This PR simply detects if the linker supports the `-relocateable` flag and only adds `-r` when that flag is supported.

Does that clear things up?

angerman commented:
This PR makes cabal agnostic to the linker. For GHCs using lld it will not try to create relocatabled pre-linked ghci libs. For GHCs using GNU ld (or any other linker that supports -r, potentially lld in the future) it permits the creation pre-linked ghci objects. These are significantly faster to load wirh our static rts linker in GHC.

While GHC linked to control the toolchain it ships with for native deployments. This PR makes cabal agnostic to the used toolchain, and especially prevents breakage for cross compiled GHCs when using the gun toolchain.

erikd commented:
Rebased and force pushed.

bgamari commented:
@erikd, can you please edit the MR description to provide an overview of what issue this MR addresses and how it does so?

erikd commented:
@bgamari What is an MR? I assume you do not mean PR, because you said MR twice.

Kleidukos commented:
@erikd MR is the term used by Gitlab

bgamari commented:
Yes, apologies. MR stands for "merge request" and is GitLab's terminology for "pull request".

erikd commented:
Updated.

bgamari commented on Cabal/src/Distribution/Simple/Program/Ld.hs:
```diff
@@ -104,3 +106,10 @@ combineObjectFiles verbosity lbi ld target files = do
       run invs
+
+    -- Prepend "-r" to the list if the linker supports relocatable outputs.
+    prependRelocatableFlag :: [String] -> [String]
```
I'm a bit lost here. It seems to me that this will simply attempt to call `ld` without `-r` to join objects when `-r` is not supported. This is, AFAICT, not correct and will at best result in linking (undefined reference) errors, and at worst result in an invalid (non-relocatable) object.

erikd replied:
If `ldProgram` doesn’t have `-r`, we shouldn’t even try to merge objects.

erikd replied:
Updated to call `error` if `combineObjectFiles` is called with an `ldProgram` that does not support relocatable output.

bgamari replied:
@Mikolaj is `error` the best way to fail here?

erikd replied:
I had a bit of a look, but `error` is used quite extensively elsewhere in the `Cabal` library.

andreabedini replied:
I believe `dieWithException` is the standard after #9219

erikd committed:
combineObjectFiles: Error if the linker does not support -r

The `lld` linker (commonly used on Windows) does not support
relocatable output.

Closes: https://github.com/haskell/cabal/issues/9414

erikd commented on Cabal/src/Distribution/Simple/Errors.hs:
```diff
@@ -301,6 +302,7 @@ exceptionCode e = case e of
   NoProgramFound{} -> 7620
   BadVersionDb{} -> 8038
   UnknownVersionDb{} -> 1008
+  CombineObjectFilesRelocatableUnsupported -> 4346
```
Could not find any information about how this number is chosen. Found another relocation based exception and chose an unused number nearby.

bgamari replied:
@SuganyaAK, @sclv, indeed we should make sure that this gets documented somehow. Would one of you be willing to do so?

andreabedini replied:
I think the relevant discussion is https://github.com/haskell/cabal/issues/8543

erikd commented:
@bgamari Ping? I think I need your approval before this can be merged.

angerman reviewed (approved):
🎉 no special casing, but feature detection!

bgamari commented on Cabal/src/Distribution/Simple/Errors.hs:
```diff
@@ -791,3 +793,6 @@ exceptionMessage e = case e of
       ++ " could not be determined."
+  CombineObjectFilesRelocatableUnsupported ->
+    "combineObjectFiles: Have been asked to combine object "
+      ++ "files but the configured linker does not support relocations."
```
The wording of this error seems wrong. The linker definitely does support relocations (e.g. nonrelocatable output still has dynamic relocations). It is rather that it doesn't support producing relocatable object files.

andreabedini commented:
There's four approvals now, @erikd you can add the "merge-me" label when you are ready.

Kleidukos commented on Cabal/src/Distribution/Simple/Program/Ld.hs:
```diff
@@ -67,6 +72,11 @@ combineObjectFiles verbosity lbi ld target files = do
+    Just "YES" -> pure ()
+    Just "NO" ->
+      dieWithException verbosity $ CombineObjectFilesRelocatableUnsupported
+    _other -> pure () -- This may still fail if the compiler does not support -r.
```
Sorry for barging in at the eleventh hour, this is not a blocking request but would you mind annotating this comment with `NOTE`?

erikd commented:
Removed the "mergeme" label because we may not actually be ready to merge.

erikd committed:
CI: Tweaks

* Update ghc versions used.
* Disable huge number of failing test for ghc == 9.4.* on Windows.

erikd commented:
Hope to replace this PR with https://github.com/haskell/cabal/pull/9443

erikd commented:
This is redundant now that #9443 is merged.
